README.md
[![View MatlabRobustNonlinLsq on File Exchange](https://www.mathworks.com/matlabcentral/images/matlab-file-exchange.svg)](https://de.mathworks.com/matlabcentral/fileexchange/62830-matlabrobustnonlinlsq)

# MatlabRobustNonlinLsq

MATLAB's function `robustfit()` enables to fit a polynomial to data whose residuals are not normally distributed, which is the implicit assumption of least squares (LSQ) methods. Non-normal residuals result, for example, from outliers in empirical data or noise which is not normally distributed. The function utilizes the *Iteratively Reweighted Least Squares* (IRLS) approach with data-adaptive weights.

The downside is that `robustfit()` does not support *non-linear least squares* problems which often arise if the mathematical model is non-linear in its parameters. The function `lsqcurvefit()` and `lsqnonlin()` provide means to solve non-linear LSQ with optional parameter bounds (due to the *trust-region-reflective* algorithm) but don't support weighted LSQ for robust estimation in the presence of outliers.

This MATLAB function is intended to give the best of both worlds, i.e. combine methods of `robustfit()` and `lsqnonlin()` to accomplish robust non-linear least squares calculations.

## Dependencies

This function is tested in MATLAB R2016b but should scale to any modern MATLAB release.

The function depends on the following toolboxes:
- Statistics and Machine Learning Toolbox
- Optimization Toolbox

## Usage

Consider the following problem. We want to fit a function to data which stems from a square-root function with a scale and location parameter. The data is noisy and exhibits strong outliers. Let's see how ordinary and robust non-linear least squares perform in this situation. The example is taken from the provided demo file.

```matlab
% the model's true parameters
trueParams = [50, -0.7];

% create some noisy data using a square root function with 2 parameters and nasty noise
x = linspace(0, 20, 100);
modelFun = @(param, x) param(1)*sqrt(max(0, x + param(2)));

noise = 10*sin(x).^2 .* randn(size(x));

y = modelFun(trueParams, x) + noise;

% introduce outliers
y(11)  = -300;
y(end) = 2000;

%%%% Fit the model %%%%
% make sure the parameters are within the following range
lb = [1, -inf];
ub = [inf, 10];

% initial guess
x0 = [1, -1];

% don't show infos from lsqcurvefit()
options = optimset(@lsqcurvefit);
options.Display = 'off';

% estimate parameters using ordinary and robust LSQ
estParamsOrdinary = lsqcurvefit(modelFun, x0, x, y, lb, ub, options);
estParamsRobust   = robustlsqcurvefit(modelFun, x0, x, y, lb, ub, [], options);

figure;
hold on;
scatter(x, y, 'filled');
plot(x, modelFun(trueParams, x), 'k--', 'linewidth', 2);
plot(x, modelFun(estParamsOrdinary, x), 'linewidth', 2');
plot(x, modelFun(estParamsRobust, x), 'linewidth', 2);
hold off;
grid on;

legend(...
    {'Noisy Data', 'True Regression', 'Ordinary non-lin LSQ', 'Robust non-lin LSQ'}, ...
    'fontsize', 12 ...
    );
```

![example.png](example.png)

## License

The code is licensed under BSD 3-Clause license.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.2.4] - 2019-11-26
### Added
- Unit test is a TestCase class now
- FileExchange banner in README

## [0.2.3] - 2019-04-02
### Fixed
- the weight vector is square-rooted since the complete cost function will be squared by the MATLAB function `lsqnonlin()`

## [0.2.2] - 2017-05-07
### Fixed
- missing version infos in the function header


robustNonlinLsq_test.m
function [tests] = robustNonlinLsq_test()
%ROBUSTNONLINLSQ_TEST <purpose in one line!>
%
% Usage: [tests] = robustNonlinLsq_test()
%
%
% Author:  J.-A. Adrian (JA) <jensalrik.adrian AT gmail.com>
% Date  :  04-May-2017 22:54
%

% History:  v0.1  initial version, 04-May-2017 (JA)
%


tests = functiontests(localfunctions());
end


function setup(testCase)

end

function teardown(testCase)

end

function setupOnce(testCase)

end

function teardownOnce(testCase)

end


function testSomething(testCase)

end

function awesomeTest(testCase)

end


% End of file: robustNonlinLsq_test.m


Robustlsqcurvefit_test.m
classdef Robustlsqcurvefit_test < matlab.unittest.TestCase
    
    properties
        Unit = @robustlsqcurvefit;
        
        TrueParams;
        X;
        Y;
        Noise;
        ModelFun;
        X0
        Options;
    end
    
    
    methods (TestClassSetup)
        function setupData(testCase)
            param = [0.5, 1.25];
            
            x = linspace(0, 6, 100);
            modelFun = @(param, x) param(1)*exp(param(2) * x);
            
            noise = sin(x).^2 .* randn(size(x));
            
            y = modelFun(param, x) + noise;
            y(10) = 200;
            y(50) = 400;
            
            options = optimset(@lsqcurvefit);
            options.Display = 'off';
            
            testCase.TrueParams = param;
            testCase.X = x;
            testCase.Y = y;
            testCase.Noise = noise;
            testCase.ModelFun = modelFun;
            testCase.X0 = [1, 1];
            testCase.Options = options;
        end
    end
    
    
    methods (Test)
        function testSimpleCall(testCase)
            x = testCase.X;
            y = testCase.Y;
            fun = testCase.ModelFun;
            x0 = testCase.X0;
            options = testCase.Options;
            
            outputs = testCase.Unit(fun, x0, x, y, [], [], [], options);
            testCase.verifyNumElements(outputs, 2);
            testCase.verifyLessThan(norm(outputs - testCase.TrueParams).^2, 1e-4);
        end
        
        
        function testFullCall(testCase)
            x = testCase.X;
            y = testCase.Y;
            fun = testCase.ModelFun;
            x0 = testCase.X0;
            options = testCase.Options;
            
            lb = [0.1, 1];
            ub = [  1, 2];
            
            weightMethod = 'bisquare';
            
            outputs = testCase.Unit(fun, x0, x, y, lb, ub, weightMethod, options);
            testCase.verifyNumElements(outputs, 2);
            testCase.verifyLessThan(norm(outputs - testCase.TrueParams).^2, 1e-4);
        end
        
        
        function testInputValidation(testCase)
            x = testCase.X;
            y = testCase.Y;
            fun = testCase.ModelFun;
            x0 = testCase.X0;
            options = testCase.Options;
            
            % Wrong WeightMethod Input
            actual = @() testCase.Unit(fun, x0, x, y, [], [], 'foo', options);
            testCase.verifyError(actual, 'MATLAB:unrecognizedStringChoice');
            
            % Wrong lb Input
            actual = @() testCase.Unit(fun, x0, x, y, 'foo', [], [], options);
            testCase.verifyError(actual, 'MATLAB:robustlsqcurvefit:invalidType');
            
            % Correct Input
            actual = @() testCase.Unit(fun, x0, x, y, [], [], 'ols', options);
            testCase.verifyWarningFree(actual);
        end
        
        
        function testPrecision(testCase)
            x = testCase.X;
            y = testCase.Y;
            fun = testCase.ModelFun;
            x0 = testCase.X0;
            options = testCase.Options;
            
            est = testCase.Unit(fun, x0, x, y, [], [], [], options);
            testCase.verifyLessThan(norm(est - testCase.TrueParams).^2, 1e-4);
        end
        
        
        function canUseLessParameters(testCase)
            x = testCase.X;
            y = testCase.Y;
            fun = testCase.ModelFun;
            x0 = testCase.X0; 
            
            outputs = testCase.Unit(fun, x0, x, y);
            testCase.verifyNumElements(outputs, 2);
        end
    end
end



robustlsqcurvefit_test.m
function [tests] = robustlsqcurvefit_test()
%ROBUSTLSQCURVEFIT_TEST <purpose in one line!>
%
% Usage: [tests] = robustlsqcurvefit_test()
%
%
% Author:  J.-A. Adrian (JA) <jensalrik.adrian AT gmail.com>
% Date  :  05-May-2017 11:27
%

% History:  v0.1.0  initial version, 05-May-2017 (JA)
%           v0.2.0  implement new unit test functions, 07-May-2017 (JA)
%


tests = functiontests(localfunctions());
end


function setupOnce(testCase)
param = [0.5, 1.25];

x = linspace(0, 6, 100);
modelFun = @(param, x) param(1)*exp(param(2) * x);

noise = sin(x).^2 .* randn(size(x));

y = modelFun(param, x) + noise;
y(10) = 200;
y(50) = 400;

options = optimset(@lsqcurvefit);
options.Display = 'off';

testCase.TestData.trueParams = param;
testCase.TestData.x = x;
testCase.TestData.y = y;
testCase.TestData.noise = noise;
testCase.TestData.modelFun = modelFun;
testCase.TestData.x0 = [1, 1];
testCase.TestData.options = options;
end



function testSimpleCall(testCase)
x = testCase.TestData.x;
y = testCase.TestData.y;
fun = testCase.TestData.modelFun;

x0 = testCase.TestData.x0;

options = testCase.TestData.options;

robustlsqcurvefit(fun, x0, x, y, [], [], [], options);
end

function testFullCall(testCase)
x = testCase.TestData.x;
y = testCase.TestData.y;
fun = testCase.TestData.modelFun;

x0 = testCase.TestData.x0;

options = testCase.TestData.options;

lb = [0.1, 1];
ub = [  1, 2];

weightMethod = 'bisquare';

outputs = cell(7, 1); %#ok<PREALL>
[outputs] = robustlsqcurvefit(fun, x0, x, y, lb, ub, weightMethod, options); %#ok<*NASGU>
end

function testInputValidation(testCase)
x = testCase.TestData.x;
y = testCase.TestData.y;
fun = testCase.TestData.modelFun;

x0 = testCase.TestData.x0;

options = testCase.TestData.options;

%% Wrong WeightMethod Input
actual = @() robustlsqcurvefit(fun, x0, x, y, [], [], 'foo', options);

testCase.verifyError(actual, 'MATLAB:unrecognizedStringChoice');

%% Wrong lb Input
actual = @() robustlsqcurvefit(fun, x0, x, y, 'foo', [], [], options);

testCase.verifyError(actual, 'MATLAB:robustlsqcurvefit:invalidType');

%% Correct Input
actual = @() robustlsqcurvefit(fun, x0, x, y, [], [], 'ols', options);

testCase.verifyWarningFree(actual);
end

function testPrecision(testCase)
x = testCase.TestData.x;
y = testCase.TestData.y;
fun = testCase.TestData.modelFun;

x0 = testCase.TestData.x0;

options = testCase.TestData.options;

est = robustlsqcurvefit(fun, x0, x, y, [], [], [], options); %#ok<*NASGU>

testCase.verifyLessThan(norm(est - testCase.TestData.trueParams).^2, 1e-4);
end






% End of file: robustlsqcurvefit_test.m


robustlsqcurvefit.m
function [varargout] = robustlsqcurvefit(fun, x0, xdata, ydata, lb, ub, weightMethod, options)
%ROBUSTLSQCURVEFIT solves robust non-linear least squares problems.
% -------------------------------------------------------------------------
% This function implements iteratively reweighted least squares (IRLS)
% using a non-linear least squares approach. Thus, it combines the strength
% of robust fitting, like known from MATLAB's robustfit() function, and the
% possibility to fit data to models which are non-linear in their
% parameters, like known from MATLAB's lsqcurvefit() function (which uses
% lsqnonlin() internally but provides a simpler interface for data
% fitting). This combination is, yet, not provided by functions in any of
% MATLAB's toolboxes.
% 
% The function computes weights iteratively to solve weighted non-linear
% least squares. The latter part is done by calling lsqnonlin().
% 
% This function has dependencies to the following toolboxes:
%   - Statistics and Machine Learning Toolbox
%   - Optimization Toolbox
% 
%
% Usage: [varargout] = robustlsqcurvefit(fun, x0, xdata, ydata)
%        [varargout] = robustlsqcurvefit(fun, x0, xdata, ydata, lb, ub)
%        [varargout] = robustlsqcurvefit(fun, x0, xdata, ydata, lb, ub, weightMethod)
%        [varargout] = robustlsqcurvefit(fun, x0, xdata, ydata, lb, ub, weightMethod, options)
%
%   Input:   ---------
%           fun - function handle to the model function, i.e. handle to the
%                 function which is likely to produce the data observed in
%                 y. This follows the definition from lsqcurvefit() [NOT
%                 the one from lsqnon()]. Consult documentation of
%                 lsqcurvefit() for details.
%           x0 - initial guess of the true parameters. Consult
%                documentation of lsqcurvefit() for details.
%           xdata - x values corresponding to the observed data in
%                   'ydata'. Consult documentation of lsqcurvefit() for
%                   details.
%           ydata - observed data to which the function 'fun' will be
%                   fitted by optimizing its parameters. Consult
%                   documentation of lsqcurvefit() for details. 
%           lb - lower bound on the design variables, X, so that the
%                solution is in the range lb <= X <= ub. Consult
%                documentation of lsqcurvefit() for details.
%                [default: lb = []]
%           ub - upper bound on the design variables, X, so that the
%                solution is in the range lb <= X <= ub. Consult
%                documentation of lsqcurvefit() for details.
%                [default: ub = []]
%           weightMethod - string defining the data weight design function.
%                          The designs are adapted to MATLAB's robustfit()
%                          and can be one of the following:
%                               - 'andrews'
%                               - 'bisquare'
%                               - 'cauchy'
%                               - 'fair'
%                               - 'huber'
%                               - 'logistic'
%                               - 'ols'
%                               - 'talwar'
%                               - 'welsch'
%                           for further information consult the
%                           documentation of robustfit().
%                           [default: weightMethod = 'bisquare']
%           options - option struct which can be created by calling
%                     optimset(@lsqcurvefit) which is directly passed to
%                     the function lsqcurvefit(). Use this to steer
%                     optimization behaviour of lsqcurvefit(). Consult
%                     documentation of lsqcurvefit() for details.
%                     [default: options = optimset(@lsqcurvefit)]
%           
%
%  Output:   ---------
%           varargout - the exact outputs of the function lsqcurvefit().
%                           - x
%                           - resnorm
%                           - residual
%                           - exitflag
%                           - output
%                           - lambda
%                           - jacobian
%                       Consult documentation of lsqcurvefit() for details.
%
%
% Author:  J.-A. Adrian (JA) <jensalrik.adrian AT gmail.com>
% Date  :  05-May-2017 11:27
%

% History:  v0.1.0  initial version, 05-May-2017 (JA)
%           v0.2.0  fix bugs, update documentation, 07-May-2017 (JA)
%           v0.2.1  update documentation, 07-May-2017 (JA)
%           v0.2.2  update version info, 07-May-2017 (JA)
%           v0.2.3  fix missing sqrt-ing of the weights, 02-Apr-2019 (JA)
%           v0.2.4  include MathWorks FileExchange banner in README, 26-Nov-2019 (JA)
%


if nargin < 8 || isempty(options)
    options = optimset(@lsqcurvefit);
end
if nargin < 7 || isempty(weightMethod)
    weightMethod = 'bisquare';
end
if nargin < 6
    ub = [];
end
if nargin < 5
    lb = [];
end
if nargin < 4
    help(mfilename)
end

[weightMethod] = ...
    validateInputArguments(fun, x0, xdata, ydata, lb, ub, weightMethod, options);

convergenceThreshold = 1e-6;

varargout = cell(max(nargout, 1), 1);

xdata = xdata(:);
ydata = ydata(:);

[weightFunction, tuningConstant] = weightFunAndConstant(weightMethod);

hasConverged     = false;
previousEstimate = inf(size(x0));
weights          = ones(size(xdata));
iterationCounter = 1;
while ~hasConverged && iterationCounter < options.MaxIter
    %%% weighted LSQ
    % define the cost function which will be squared and summed by lsqnonlin(). Due to the latter
    % point, the weights have to be square-rooted here.
    weightedFun = ...
        @(params) (fun(params, xdata) - ydata) .* sqrt(weights);
    
    varargout{:} = lsqnonlin(weightedFun, x0, lb, ub, options);
    
    thisEstimate = varargout{1};
    hasConverged = norm(thisEstimate - previousEstimate)^2 < convergenceThreshold;
    
    %%% update weights
    residuals = ydata - fun(thisEstimate, xdata);
    residuals = residuals(:);
    
    residualLeverages = leverage(residuals);
    robustVar         = mad(residuals, 1);
    
    r = residuals ./ (tuningConstant * robustVar * sqrt(1 - residualLeverages));
    
    weights = weightFunction(r);
    previousEstimate = thisEstimate;
    iterationCounter = iterationCounter + 1;
end
end


function [weightMethod] = validateInputArguments(...
    fun, ...
    x0, ...
    xdata, ...
    ydata, ...
    lb, ...
    ub, ...
    weightMethod, ...
    options ...
    )

validateattributes(...
    fun, ...
    {'function_handle'}, ...
    {'scalar', 'nonempty'}, ...
    mfilename, ...
    'fun', ...
    1 ...
    );
assert(...
    nargin(fun) == 2, ...
    ['No. of input arguments of ''fun'' do not match the desired no. of 2 ', ...
    '(parameters and x-values)'] ...
    );

validateattributes(...
    x0, ...
    {'numeric'}, ...
    {'vector', 'nonempty', 'finite'}, ...
    mfilename, ...
    'x0', ...
    2 ...
    );

validateattributes(...
    xdata, ...
    {'numeric'}, ...
    {'vector', 'nonempty', 'finite'}, ...
    mfilename, ...
    'xdata', ...
    3 ...
    );

validateattributes(...
    ydata, ...
    {'numeric'}, ...
    {'vector', 'nonempty', 'finite'}, ...
    mfilename, ...
    'ydata', ...
    4 ...
    );

validateattributes(...
    lb, ...
    {'numeric'}, ...
    {}, ...
    mfilename, ...
    'lb', ...
    5 ...
    );

validateattributes(...
    ub, ...
    {'numeric'}, ...
    {}, ...
    mfilename, ...
    'ub', ...
    6 ...
    );

validateattributes(...
    weightMethod, ...
    {'char'}, ...
    {'nonempty'}, ...
    mfilename, ...
    'weightMethod', ...
    7 ...
    );

validateattributes(...
    options, ...
    {'numeric', 'struct'}, ...
    {}, ...
    mfilename, ...
    'options', ...
    8 ...
    );

weightMethod = validatestring(...
    weightMethod, ...
    {'bisquare', 'andrews', 'cauchy', 'fair', 'huber', 'logistic', 'ols', 'talwar', 'welsch'} ...
    );
end

function [weightFun, tuningConstant] = weightFunAndConstant(method)
switch lower(method)
    case 'bisquare'
        weightFun      = @(r) (abs(r) < 1) .* (1 - r.^2).^2;
        tuningConstant = 4.685;
        
    case 'andrews'
        weightFun      = @(r) (abs(r) < pi) .* sin(r) ./ r;
        tuningConstant = 1.339;
        
    case 'cauchy'
        weightFun      = @(r) 1 ./ (1 + r.^2);
        tuningConstant = 2.385;
        
    case 'fair'
        weightFun      = @(r) 1 ./ (1 + abs(r));
        tuningConstant = 1.4;
        
    case 'huber'
        weightFun      = @(r) 1 ./ max(1, abs(r));
        tuningConstant = 1.345;
        
    case 'logistic'
        weightFun      = @(r) tanh(r) ./ r;
        tuningConstant = 1.205;
        
    case 'ols'
        weightFun      = @(r) ones(size(r));
        tuningConstant = 1;
        
    case 'talwar'
        weightFun      = @(r) 1 * (abs(r) < 1);
        tuningConstant = 2.795;
        
    case 'welsch'
        weightFun      = @(r) exp(-(r.^2));
        tuningConstant = 2.985;
        
end
end



robustnonlinlsqcurvefit.m
function [varargout] = robustnonlinlsqcurvefit(fun, x0, xdata, ydata, lb, ub, options)
%ROBUSTNONLINLSQCURVEFIT <purpose in one line!>
% -------------------------------------------------------------------------
% <Detailed description of the function>
%
% Usage: [y] = robustnonlinlsqcurvefit(input)
%
%   Input:   ---------
%
%  Output:   ---------
%
%
% Author:  J.-A. Adrian (JA) <jensalrik.adrian AT gmail.com>
% Date  :  05-May-2017 11:27
%

% History:  v0.1  initial version, 05-May-2017 (JA)
%


if nargin < 7 || isempty(options)
    options = optimset('lsqnonlin');
end
if nargin < 6
    ub = [];
end
if nargin < 5
    lb = [];
end
if nargin < 4
    help(mfilename)
end


varargout = cell(nargout, 1);

hasConverged     = false;
previousEstimate = inf(size(x0));
weights          = ones(size(xdata));
while ~hasConverged
    %%% weighted LSQ
    weightedFun = ...
        @(params) (fun(params, xdata) - ydata) .* weights;
    
    [varargout{:}] = ...
        lsqnonlin(weightedFun, x0, lb, ub, options);
    
    thisEstimate = varargout{1};
    hasConverged = norm(thisEstimate - previousEstimate) > options.TolFun;
    
    %%% update weights
    residuals = ydata - weightedFun(thisEstimate);
    residuals = residuals(:);
    
    %TODO: correct leverages
    theseLeverages = leverage(residuals);
    tuningConstant = 4.685;
    robustVar      = mad(residuals, 1);
    
    r = residuals ./ (tuningConstant * robustVar * sqrt(1 - theseLeverages));
    
    weights = (abs(r) < 1) .* (1 - r.^2).^2;
end



% End of file: robustnonlinlsqcurvefit.m


robustlsqcurvefit_demo.m
% Demonstrate robustlsqcurvefit() in some applications
%
% Author:  J.-A. Adrian (JA) <jensalrik.adrian AT gmail.com>
% Date  :  05-May-2017 22:46:35
%

clear;
close all;

%% Easy Linear Model With Noise and Outliers
x = (0:10)';
rng default; % For reproducibility
y = -2*x + 10 + randn(size(x));

% introduce outlier
y(11) = 0;

modelFun = @(params, x) params(1)*x + params(2);
x0       = [-1, 1];

options = optimset(@lsqcurvefit);
options.Display = 'off';

estParamsOrdinary  = polyfit(x, y, 1);
estParamsRobust    = robustlsqcurvefit(modelFun, x0, x, y, [], [], [], options);
estLinearRobustFit = robustfit(x, y);

figure;
hold on;
scatter(x, y, 'filled');
plot(x, modelFun([-2, 10], x), 'k--', 'linewidth', 2);
plot(x, modelFun(estParamsOrdinary, x), 'linewidth', 2);
plot(x, modelFun(estParamsRobust, x), 'linewidth', 2);
hold off;
grid on;

legend(...
    {'Noisy Data', 'True Regression', 'Ordinary non-lin LSQ', 'Robust non-lin LSQ'}, ...
    'fontsize', 12 ...
    );

%% Exponential Model With Noise and Outliers
trueParams = [0.5, 1.25];

x = linspace(0, 6, 100);
modelFun = @(param, x) param(1)*exp(param(2) * x);

noise = 20*sin(x).^2 .* randn(size(x));

y = modelFun(trueParams, x) + noise;

% introduce outliers
y(1)  = -400;
y(end) = 2000;

x0 = [0.3, 2];

options = optimset(@lsqcurvefit);
options.Display = 'off';

estParamsOrdinary = lsqcurvefit(modelFun, x0, x, y, [], [], options);
estParamsRobust   = robustlsqcurvefit(modelFun, x0, x, y, [], [], [], options);

figure;
hold on;
scatter(x, y, 'filled');
plot(x, modelFun(trueParams, x), 'k--', 'linewidth', 2);
plot(x, modelFun(estParamsOrdinary, x), 'linewidth', 2);
plot(x, modelFun(estParamsRobust, x), 'linewidth', 2);
hold off;
grid on;

legend(...
    {'Noisy Data', 'True Regression', 'Ordinary non-lin LSQ', 'Robust non-lin LSQ'}, ...
    'fontsize', 12 ...
    );

%% Square-Root Function Fit incl. Lower and Upper Parameter Bounds
% the model's true parameters
trueParams = [50, -0.7];

% create some noisy data using a square root function with 2 parameters and nasty noise
x = linspace(0, 20, 100);
modelFun = @(param, x) param(1)*sqrt(max(0, x + param(2)));

noise = 10*sin(x).^2 .* randn(size(x));

y = modelFun(trueParams, x) + noise;

% introduce outliers
y(11)  = -300;
y(end) = 2000;

% make sure the parameters are within the following range
lb = [1, -inf];
ub = [inf, 10];

% initial guess
x0 = [1, -1];

% don't show infos from lsqcurvefit()
options = optimset(@lsqcurvefit);
options.Display = 'off';

% estimate parameters using ordinary and robust LSQ
estParamsOrdinary = lsqcurvefit(modelFun, x0, x, y, lb, ub, options);
estParamsRobust   = robustlsqcurvefit(modelFun, x0, x, y, lb, ub, [], options);

figure;
hold on;
scatter(x, y, 'filled');
plot(x, modelFun(trueParams, x), 'k--', 'linewidth', 2);
plot(x, modelFun(estParamsOrdinary, x), 'linewidth', 2');
plot(x, modelFun(estParamsRobust, x), 'linewidth', 2);
hold off;
grid on;

title('Square-Root Fit');
xlabel('x data');
ylabel('y data');

legend(...
    {'Noisy Data', 'True Regression', 'Ordinary non-lin LSQ', 'Robust non-lin LSQ'}, ...
    'fontsize', 12 ...
    );



% End of file: robustlsqcurvefit_demo.m
